In [15]:
import matplotlib.pyplot as plt
import numpy as np
import seaborn as sns
import pandas as pd
import statsmodels.api as sm
import statsmodels.tsa.api as smt
import statsmodels.formula.api as smf
import pylab as py
import warnings
from sklearn import linear_model
from sqlalchemy import create_engine
from statsmodels.stats.outliers_influence import variance_inflation_factor

warnings.filterwarnings(action="ignore")
postgres_user = 'dsbc_student'
postgres_pw = '7*.8G9QH21'
postgres_host = '142.93.121.174'
postgres_port = '5432'
postgres_db = 'houseprices'

In [16]:
engine = create_engine('postgresql://{}:{}@{}:{}/{}'.format(
    postgres_user, postgres_pw, postgres_host, postgres_port, postgres_db))
houseprices_df = pd.read_sql_query('select * from houseprices',con=engine)

# no need for an open connection, as we're only doing a single query
engine.dispose()

# Run your house prices model again and assess the goodness of fit of your model using F-test, R-squared, adjusted R-squared, AIC and BIC. Do you think your model is satisfactory? If so, why?

In [17]:
# x = features
Y = houseprices_df[['saleprice']]
# target variable
X = houseprices_df[['overallqual', 'grlivarea', 'garagecars', 'garagearea', 'totalbsmtsf','firstflrsf']]

X_const = houseprices_df[['overallqual', 'grlivarea', 'garagecars', 'garagearea', 'totalbsmtsf','firstflrsf']]

# We create a LinearRegression model object
# from scikit-learn's linear_model module.
lrm = linear_model.LinearRegression()

# fit method estimates the coefficients using OLS
lrm.fit(X, Y)

# Inspect the results.
print('\nCoefficients: \n', lrm.coef_)
print('\nIntercept: \n', lrm.intercept_)


Coefficients: 
 [[2.39970394e+04 4.31228864e+01 1.45151932e+04 1.56639341e+01
  2.43907676e+01 1.11859135e+01]]

Intercept: 
 [-102650.90069029]


In [18]:
X = sm.add_constant(X)

results = sm.OLS(Y, X).fit()

results.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:              saleprice   R-squared:                       0.762
Model:                            OLS   Adj. R-squared:                  0.761
Method:                 Least Squares   F-statistic:                     775.0
Date:                Thu, 06 Feb 2020   Prob (F-statistic):               0.00
Time:                        18:59:32   Log-Likelihood:                -17496.
No. Observations:                1460   AIC:                         3.501e+04
Df Residuals:                    1453   BIC:                         3.504e+04
Df Model:                           6                                         
Covariance Type:            nonrobust                                         
===============================================================================
                  coef    std err          t      P>|t|      [0.025      0.975]
-------------------------------------------------------------------------------
const       -1.027e+05   4903.994    -20.932      0.000   -1.12e+05    -9.3e+04
overallqual    2.4e+04   1083.393     22.150      0.000    2.19e+04    2.61e+04
grlivarea      43.1229      2.679     16.095      0.000      37.867      48.379
garagecars   1.452e+04   3018.621      4.809      0.000    8593.872    2.04e+04
garagearea     15.6639     10.475      1.495      0.135      -4.884      36.212
totalbsmtsf    24.3908      4.318      5.649      0.000      15.921      32.860
firstflrsf     11.1859      5.032      2.223      0.026       1.315      21.057
==============================================================================
Omnibus:                      431.781   Durbin-Watson:                   1.975
Prob(Omnibus):                  0.000   Jarque-Bera (JB):            39208.253
Skew:                          -0.313   Prob(JB):                         0.00
Kurtosis:                      28.380   Cond. No.                     1.14e+04
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 1.14e+04. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

F statistic and its associated p-value are 775.0 and 0.00.
R-squared and adjusted R-squared of the model are 0.762 and 0.761.
AIC and BIC of the model are 35010 and 35040.

# In order to improve the goodness of fit of your model, try different model specifications by adding or removing some variables.

In [44]:
houseprices_df = pd.concat([houseprices_df,pd.get_dummies(houseprices_df.mszoning, prefix="mszoning", drop_first=True)], axis=1)
houseprices_df = pd.concat([houseprices_df,pd.get_dummies(houseprices_df.street, prefix="street", drop_first=True)], axis=1)
dummy_column_names = list(pd.get_dummies(houseprices_df.mszoning, prefix="mszoning", drop_first=True).columns)
dummy_column_names = dummy_column_names + list(pd.get_dummies(house_prices_df.street, prefix="street", drop_first=True).columns)

In [52]:
# x = features
Y = houseprices_df[['saleprice']]
# target variable
X = houseprices_df[['overallqual', 'grlivarea', 'garagecars', 'garagearea',
                    'totalbsmtsf','firstflrsf','totrmsabvgrd','yearbuilt'] + dummy_column_names]

# We create a LinearRegression model object
# from scikit-learn's linear_model module.
lrm = linear_model.LinearRegression()

# fit method estimates the coefficients using OLS
lrm.fit(X, Y)

# Inspect the results.
print('\nCoefficients: \n', lrm.coef_)
print('\nIntercept: \n', lrm.intercept_)


Coefficients: 
 [[ 2.12799525e+04  5.02630269e+01  1.02913469e+04  1.66995957e+01
   2.06369726e+01  9.22120904e+00 -9.25849291e+02  2.40701169e+02
   6.79954804e+03  6.79954804e+03  4.25466432e+03  4.25466432e+03
   1.02402749e+04  1.02402749e+04  3.30170800e+03  3.30170800e+03
   9.43717925e+02  9.43717925e+02]]

Intercept: 
 [-571833.64060299]


In [53]:
X = sm.add_constant(X)

results = sm.OLS(Y, X).fit()

results.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:              saleprice   R-squared:                       0.773
Model:                            OLS   Adj. R-squared:                  0.771
Method:                 Least Squares   F-statistic:                     379.4
Date:                Thu, 06 Feb 2020   Prob (F-statistic):               0.00
Time:                        20:05:25   Log-Likelihood:                -17461.
No. Observations:                1460   AIC:                         3.495e+04
Df Residuals:                    1446   BIC:                         3.502e+04
Df Model:                          13                                         
Covariance Type:            nonrobust                                         
================================================================================
                   coef    std err          t      P>|t|      [0.025      0.975]
--------------------------------------------------------------------------------
const        -5.718e+05   9.31e+04     -6.143      0.000   -7.54e+05   -3.89e+05
overallqual   2.128e+04   1173.034     18.141      0.000     1.9e+04    2.36e+04
grlivarea       50.2630      4.105     12.246      0.000      42.211      58.315
garagecars    1.029e+04   3054.930      3.369      0.001    4298.778    1.63e+04
garagearea      16.6996     10.380      1.609      0.108      -3.661      37.060
totalbsmtsf     20.6370      4.290      4.811      0.000      12.223      29.051
firstflrsf       9.2212      5.031      1.833      0.067      -0.648      19.091
totrmsabvgrd  -925.8493   1122.397     -0.825      0.410   -3127.551    1275.852
yearbuilt      240.7012     48.125      5.002      0.000     146.299     335.103
mszoning_FV   6799.5480   6876.386      0.989      0.323   -6689.212    2.03e+04
mszoning_FV   6799.5480   6876.386      0.989      0.323   -6689.212    2.03e+04
mszoning_RH   4254.6643   7874.194      0.540      0.589   -1.12e+04    1.97e+04
mszoning_RH   4254.6643   7874.194      0.540      0.589   -1.12e+04    1.97e+04
mszoning_RL   1.024e+04   6355.361      1.611      0.107   -2226.439    2.27e+04
mszoning_RL   1.024e+04   6355.361      1.611      0.107   -2226.439    2.27e+04
mszoning_RM   3301.7080   6397.635      0.516      0.606   -9247.930    1.59e+04
mszoning_RM   3301.7080   6397.635      0.516      0.606   -9247.930    1.59e+04
street_Pave    943.7179   8078.679      0.117      0.907   -1.49e+04    1.68e+04
street_Pave    943.7179   8078.679      0.117      0.907   -1.49e+04    1.68e+04
==============================================================================
Omnibus:                      436.275   Durbin-Watson:                   1.988
Prob(Omnibus):                  0.000   Jarque-Bera (JB):            48459.392
Skew:                          -0.219   Prob(JB):                         0.00
Kurtosis:                      31.221   Cond. No.                     2.45e+20
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The smallest eigenvalue is 2.22e-31. This might indicate that there are
strong multicollinearity problems or that the design matrix is singular.
"""

R-squared and adjusted R-squared of the model are 0.773 and 0.771 respectively. 
F statistic and its associated p-value are 379.4 and 0.00. This also indicates that the second model is better than the first one.
AIC and BIC of the model are 34950 and 35020 . These values are lower than the ones of the first model. The second model is better than the first model because of the R-squared and adjusted R-squared is larger in the second model.